### PR TITLE
refactor: introduce `Symbol.TypeVarSym`

### LINKED ISSUE #2510: Type.Var has no way to distinguish between the declaration site of the variable and the variable itself.
Thus, for syntax highlighting, we cannot highlight both of these occurrences:

image

### PR DESCRIPTION
**Pull request by mlutze**
related to #2510

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/ast/Symbol.scala [hunk 1]
```diff
@@ -214,6 +219,24 @@ object Symbol {
     override def toString: String = text + Flix.Delimiter + id
   }
 
+  // MATT docs
+  // MATT include kind?
+  // MATT rigidity?
+  // MATT toString
+  final class TypeVarSym(val id: Int, val text: Option[String], val loc: SourceLocation) extends Sourceable with Locatable {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Symbol.scala:226]: Overall I think this is the right approach.

File: main/src/ca/uwaterloo/flix/language/ast/Type.scala [hunk 2]
```diff
@@ -372,56 +372,56 @@ object Type {
     * A type variable.
     */
   @IntroducedBy(Kinder.getClass)
-  case class KindedVar(id: Int, kind: Kind, loc: SourceLocation, rigidity: Rigidity = Rigidity.Flexible, text: Option[String] = None) extends Type with Var with BaseType with Ordered[Type.KindedVar] {
+  case class KindedVar(sym: Symbol.TypeVarSym, kind: Kind, loc: SourceLocation, rigidity: Rigidity = Rigidity.Flexible, text: Option[String] = None) extends Type with Var with BaseType with Ordered[Type.KindedVar] {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Type.scala:375]: Looks good. I guess the really difficult question is how much of this data to leave here and how much to put into Symbol.TypeVar.

**Comment by mlutze** [comment]
Marking as RFR. This ends up only being a refactor (no behavioral changes) as the changes that do actually separate declaration and use locations have turned out to be quite substantial and merit some scrutiny/discussion.

File: main/src/ca/uwaterloo/flix/language/ast/Type.scala [hunk 3]
```diff
@@ -372,56 +385,63 @@ object Type {
     * A type variable.
     */
   @IntroducedBy(Kinder.getClass)
-  case class KindedVar(id: Int, kind: Kind, loc: SourceLocation, rigidity: Rigidity = Rigidity.Flexible, text: Option[String] = None) extends Type with Var with BaseType with Ordered[Type.KindedVar] {
+  case class KindedVar(sym: Symbol.KindedTypeVarSym, loc: SourceLocation) extends Type with Var with BaseType with Ordered[Type.KindedVar] {
```
[the remark this answers is not in the document]
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/Type.scala:388]: Possibly. But I think if we do that, we might end up getting a lot of `tvar.sym.asKinded`.

File: main/test/ca/uwaterloo/flix/language/fmt/TestFormatType.scala [hunk 4]
```diff
@@ -24,6 +24,8 @@ class TestFormatType extends FunSuite with TestUtils {
 
   val loc: SourceLocation = SourceLocation.Unknown
 
+  private def mkTvarSym(id: Int, kind: Kind, rigidity: Rigidity, text: Option[String]): Symbol.KindedTypeVarSym = new Symbol.KindedTypeVarSym(id, text, kind, rigidity, loc)
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/language/fmt/TestFormatType.scala:27]: This seems to just purely forward, so I don't think its needed.

File: main/test/ca/uwaterloo/flix/language/phase/unification/TestUnification.scala [hunk 5]
```diff
@@ -29,6 +29,8 @@ class TestUnification extends FunSuite with TestUtils {
 
   val loc: SourceLocation = SourceLocation.Unknown
 
+  private def mkTvarSym(id: Int, kind: Kind, rigidity: Rigidity, text: Option[String]): Symbol.KindedTypeVarSym = new Symbol.KindedTypeVarSym(id, text, kind, rigidity, loc)
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/language/phase/unification/TestUnification.scala:32]: Same comment.

File: main/src/ca/uwaterloo/flix/language/ast/Symbol.scala [hunk 6]
```diff
@@ -209,6 +223,70 @@ object Symbol {
...
+
+    /**
+      * Returns a string representation of the symbol.
+      */
+    override def toString: String = text.getOrElse("tvar") + Flix.Delimiter + id
+  }
+
+  /**
+    * Kinded type variable symbol.
+    *
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Symbol.scala:259]: Extra newline

File: main/src/ca/uwaterloo/flix/language/ast/Symbol.scala [hunk 7]
```diff
@@ -209,6 +223,70 @@ object Symbol {
...
+
+    override def withText(newText: Option[String]): KindedTypeVarSym = new KindedTypeVarSym(id, newText, kind, rigidity, loc)
+
+    override def withRigidity(newRigidity: Rigidity): KindedTypeVarSym = new KindedTypeVarSym(id, text, kind, newRigidity, loc)
+
+  }
+
+  /**
+    * Unkinded type variable symbol.
+    *
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Symbol.scala:276]: Extra newline

**Comment by magnus-madsen** [comment]
I think this is ready, I only saw some minor things.

**Pull request merged**